go/tools/gopackagesdriver/pkgjson/main.go

```diff
@@ -0,0 +1,228 @@
+func processCgoFiles(cgoOutDir string, p *pkgJson, pathResolver func(string) string) error {
+	cf, err := filterCgoSourceFiles(p.CompiledGoFiles, pathResolver)
+	if err != nil {
+		return err
+	}
+	p.CompiledGoFiles = cf.goSourceFiles
+	if cgoOutDir == "" {
+		return nil
+	}
+	cgoSrcs, err := cgoGoSrcs(cgoOutDir, cf.cgoSourceFiles, pathResolver)
+	if err != nil {
+		return err
+	}
+	p.CompiledGoFiles = append(p.CompiledGoFiles, cgoSrcs...)
+	return nil
+}
+
+type cgoFiles struct {
+	cgoSourceFiles []string
+	goSourceFiles []string
+}
+
+func filterCgoSourceFiles(compiledGoFiles []string, resolvePath func(string) string) (cgoFiles, error) {
+	filtered := make([]string, 0, len(compiledGoFiles))
+	var cgoSourceFiles []string
+	fset := token.NewFileSet()
+	for _, file := range compiledGoFiles {
+		// Note: We check resolved file paths to get a path relative to the execroot
+		// but we add the unresolved file path to the package json as the gopackagesdriver
+		// uses the special bazel prefixes to resolve to absolute paths.
+		resolvedPath := resolvePath(file)
+		f, err := parser.ParseFile(fset, resolvedPath, nil, parser.ImportsOnly)
+		if err != nil {
+			return cgoFiles{}, err
+		}
+		var skip bool
+		for _, rawImport := range f.Imports {
+			imp, err := strconv.Unquote(rawImport.Path.Value)
+			if err != nil {
+				continue
+			}
+			if imp == "C" {
+				skip = true
+				break
+			}
+		}
+		if skip {
+			// Skip Cgo preprocessed files.
+			cgoSourceFiles = append(cgoSourceFiles, file)
+		} else {
+			filtered = append(filtered, file)
+		}
+	}
+	return cgoFiles{
+		cgoSourceFiles: cgoSourceFiles,
+		goSourceFiles:  filtered,
+	}, nil
+}
+
+func cgoGoSrcs(cgoGeneratedDir string, cgoSourceFiles []string, resolvePath func(string) string) ([]string, error) {
+	var cgoGeneratedFiles []string
+	// Only include cgo generated files if they were generated by Bazel, using
+	// the _cgo_gotypes.go file name as a marker for cgo files as it's always
+	// present if the package contains cgo generated code.
+	//
+	// Note: We check resolved file paths to get a path relative to the execroot
+	// but we add the unresolved file path to the package json as the gopackagesdriver
+	// uses the special bazel prefixes to resolve to absolute paths.
+	resolvedCgoGeneratedDir := resolvePath(cgoGeneratedDir)
+	path := filepath.Join(resolvedCgoGeneratedDir, "_cgo_gotypes.go")
+	if _, err := os.Stat(path); err != nil && os.IsNotExist(err) {
+		return nil, nil
+	}
+	cgoGeneratedFiles = append(cgoGeneratedFiles, filepath.Join(cgoGeneratedDir, "_cgo_gotypes.go"))
+	cgoGeneratedFiles = append(cgoGeneratedFiles, filepath.Join(cgoGeneratedDir, "_cgo_imports.go"))
+	for _, csf := range cgoSourceFiles {
+		name := strings.TrimSuffix(filepath.Base(csf), ".go")
+		name = name + ".cgo1.go"
+		resolvedPath := filepath.Join(resolvedCgoGeneratedDir, name)
+		unresolvedPath := filepath.Join(cgoGeneratedDir, name)
+		if _, err := os.Stat(resolvedPath); err != nil && os.IsNotExist(err) {
```

Comment:
Seems like this combo of "Check file exists, but I only have the \"unresolved\" path" happens a few times. Can we have a utility function?

Reply:
This can't be easily placed in a utility function as what happens as a result of the check varies in each of the check cases (e.g. return nil, append to a slice, or append after checking cgo imports). Lines of code aren't really saved by making a utility function out of a check as from the check's perspective it doesn't know if the path is resolved or not.